Crawl command job argument not working

### Affected version(s)

4.9, 4.13

### Description

Using the `contao:crawl` command with an job id (existing in the queue or not) always lead to ` [ERROR] Could not find the given job ID. ` As already discussed on [slack](https://contao.slack.com/archives/CK4J0KNDB/p1668156880869559), the problem is, that the queue feature simply is not implemented in the crawl command.

There are two possible solutions:

1. Remove the argument from the command as long as it is not implemented.
2. Implement the queue in the command (which would be the better, but requires more work)

I did not test it in contao 5, but I think the issue is still existing there.

--- a/core-bundle/src/Command/CrawlCommand.php
+++ b/core-bundle/src/Command/CrawlCommand.php
@@ -68,6 +68,7 @@ class CrawlCommand extends Command
     {
         $this
             ->addArgument('job', InputArgument::OPTIONAL, 'An optional existing job ID')
+            ->addOption('queue', 'q', InputArgument::OPTIONAL, 'Queue to use ("memory" or "doctrine")', 'memory')
             ->addOption('subscribers', 's', InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY, 'A list of subscribers to enable', $this->escargotFactory->getSubscriberNames())
             ->addOption('concurrency', 'c', InputOption::VALUE_REQUIRED, 'The number of concurrent requests that are going to be executed', '10')
             ->addOption('delay', null, InputOption::VALUE_REQUIRED, 'The number of microseconds to wait between requests (0 = throttling is disabled)', '0')
@@ -86,7 +87,6 @@ class CrawlCommand extends Command
         $io->title('Contao Crawler');
 
         $subscribers = $input->getOption('subscribers');
-        $queue = new InMemoryQueue();
         $baseUris = $this->escargotFactory->getCrawlUriCollection();
 
         if ($baseUris->containsHost('localhost')) {
@@ -94,6 +94,20 @@ class CrawlCommand extends Command
         }
 
         try {
+            switch ($input->getOption('queue')) {
+                case 'memory':
+                    $queue = new InMemoryQueue();
+                    break;
+                case 'doctrine':
+                    $queue = $this->escargotFactory->createLazyQueue();
+                    break;
+
+                default:
+                    $io->error('Only "memory" or "doctrine" are allowed for the "queue" option.');
+
+                    return 1;
+            }
+
             if ($jobId = $input->getArgument('job')) {
                 $this->escargot = $this->escargotFactory->createFromJobId($jobId, $queue, $subscribers);
             } else {
@@ -127,7 +141,12 @@ class CrawlCommand extends Command
 
         $output->writeln('');
         $output->writeln('');
-        $io->comment('Finished crawling! Find the details for each subscriber below:');
+        $io->comment(
+            sprintf(
+                '[Job ID: %s] Finished crawling! Find the details for each subscriber below:',
+                $this->escargot->getJobId()
+            )
+        );
 
         $errored = false;
 

--- a/core-bundle/src/Command/CrawlCommand.php
+++ b/core-bundle/src/Command/CrawlCommand.php
@@ -93,21 +93,21 @@ class CrawlCommand extends Command
             $io->warning('You are going to crawl localhost URIs. This is likely not desired and due to a missing domain configuration in your root page settings. You may also configure a fallback request context using "router.request_context.*" if you want to execute all CLI commands with the same request context.');
         }
 
+        switch ($input->getOption('queue')) {
+            case 'memory':
+                $queue = new InMemoryQueue();
+                break;
+            case 'doctrine':
+                $queue = $this->escargotFactory->createLazyQueue();
+                break;
+
+            default:
+                $io->error('Only "memory" or "doctrine" are allowed for the "queue" option.');
+
+                return 1;
+        }
+
         try {
-            switch ($input->getOption('queue')) {
-                case 'memory':
-                    $queue = new InMemoryQueue();
-                    break;
-                case 'doctrine':
-                    $queue = $this->escargotFactory->createLazyQueue();
-                    break;
-
-                default:
-                    $io->error('Only "memory" or "doctrine" are allowed for the "queue" option.');
-
-                    return 1;
-            }
-
             if ($jobId = $input->getArgument('job')) {
                 $this->escargot = $this->escargotFactory->createFromJobId($jobId, $queue, $subscribers);
             } else {

--- a/core-bundle/src/Command/CrawlCommand.php
+++ b/core-bundle/src/Command/CrawlCommand.php
@@ -97,6 +97,7 @@ class CrawlCommand extends Command
             case 'memory':
                 $queue = new InMemoryQueue();
                 break;
+
             case 'doctrine':
                 $queue = $this->escargotFactory->createLazyQueue();
                 break;
@@ -141,6 +142,7 @@ class CrawlCommand extends Command
 
         $output->writeln('');
         $output->writeln('');
+
         $io->comment(
             sprintf(
                 '[Job ID: %s] Finished crawling! Find the details for each subscriber below:',

--- a/core-bundle/tests/Command/CrawlCommandTest.php
+++ b/core-bundle/tests/Command/CrawlCommandTest.php
@@ -27,6 +27,7 @@ use Terminal42\Escargot\BaseUriCollection;
 use Terminal42\Escargot\Escargot;
 use Terminal42\Escargot\Exception\InvalidJobIdException;
 use Terminal42\Escargot\Queue\InMemoryQueue;
+use Terminal42\Escargot\Queue\LazyQueue;
 
 class CrawlCommandTest extends TestCase
 {
@@ -93,6 +94,57 @@ class CrawlCommandTest extends TestCase
         $this->assertSame(20, $command->getEscargot()->getRequestDelay());
         $this->assertSame(20, $command->getEscargot()->getMaxRequests());
         $this->assertSame(20, $command->getEscargot()->getMaxDepth());
+    }
+
+    public function testSubsequentCrawlCommandsWithDoctrineQueue(): void
+    {
+        $client = new MockHttpClient();
+        $queue = new InMemoryQueue();
+        $escargotFactory = $this->createMock(Factory::class);
+        $escargotFactory
+            ->expects($this->exactly(2))
+            ->method('getCrawlUriCollection')
+            ->willReturn($this->getBaseUriCollection())
+        ;
+
+        $jobId = null;
+
+        $escargotFactory
+            ->expects($this->once())
+            ->method('create')
+            ->willReturnCallback(
+                function () use ($client, $queue, &$jobId) {
+                    $escargot = Escargot::create($this->getBaseUriCollection(), $queue)->withHttpClient($client);
+                    $jobId = $escargot->getJobId();
+
+                    return $escargot;
+                }
+            )
+        ;
+
+        $escargotFactory
+            ->expects($this->exactly(2))
+            ->method('createLazyQueue')
+            ->willReturn(new LazyQueue(new InMemoryQueue(), $queue))
+        ;
+
+        $escargotFactory
+            ->expects($this->once())
+            ->method('createFromJobId')
+            ->willReturnCallback(static fn (string $jobId) => Escargot::createFromJobId($jobId, $queue)->withHttpClient($client))
+        ;
+
+        $command = new CrawlCommand($escargotFactory, new Filesystem());
+
+        $tester = new CommandTester($command);
+        $tester->execute(['-q' => 'doctrine']);
+
+        $expectCli = sprintf('[Job ID: %s]', $jobId);
+
+        $this->assertStringContainsString($expectCli, $tester->getDisplay(true));
+
+        $tester->execute(['-q' => 'doctrine', 'job' => $jobId]);
+        $this->assertStringContainsString($expectCli, $tester->getDisplay(true));
     }
 
     public function testEmitsWarningIfLocalhostIsInCollection(): void
@@ -172,6 +224,11 @@ class CrawlCommandTest extends TestCase
             $escargotFactory
                 ->expects($this->never())
                 ->method('create')
+            ;
+
+            $escargotFactory
+                ->expects($this->never())
+                ->method('createLazyQueue')
             ;
 
             return $escargotFactory;

--- a/core-bundle/tests/Command/CrawlCommandTest.php
+++ b/core-bundle/tests/Command/CrawlCommandTest.php
@@ -100,14 +100,14 @@ class CrawlCommandTest extends TestCase
     {
         $client = new MockHttpClient();
         $queue = new InMemoryQueue();
+        $jobId = null;
+
         $escargotFactory = $this->createMock(Factory::class);
         $escargotFactory
             ->expects($this->exactly(2))
             ->method('getCrawlUriCollection')
             ->willReturn($this->getBaseUriCollection())
         ;
-
-        $jobId = null;
 
         $escargotFactory
             ->expects($this->once())
@@ -144,6 +144,7 @@ class CrawlCommandTest extends TestCase
         $this->assertStringContainsString($expectCli, $tester->getDisplay(true));
 
         $tester->execute(['-q' => 'doctrine', 'job' => $jobId]);
+
         $this->assertStringContainsString($expectCli, $tester->getDisplay(true));
     }
 

